### PR TITLE
WIP: Basic support for maven in wrap depedency system

### PR DESCRIPTION
**Pull request by noverby**
Depends on: https://github.com/mesonbuild/meson/pull/3763

### CODE REVIEW

File: mesonbuild/compilers/java.py [hunk 1]
```diff
@@ -113,3 +113,11 @@ def sanity_check(self, work_dir, environment):
...
     def needs_static_linker(self):
         return False
+
+    def find_library(self, libname, env, extra_dirs):
+        for d in extra_dirs:
+            trial = os.path.join(d, libname)
+            if os.path.isfile(trial):
+                return ['-cp', trial]
+        return None
+
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/java.py]: __[Flake8]__

[W391] blank line at end of file

File: mesonbuild/wrap/wrap.py [hunk 2]
```diff
@@ -261,6 +265,31 @@ def get_svn(self, p):
...
+
+        target_dir = os.path.join(self.subdir_root, p.get('directory'))
+        if not os.path.exists(target_dir):
+            os.mkdir(target_dir)
+        shutil.copyfile(jarfile, os.path.join(target_dir, jarname))
+        with open(os.path.join(target_dir, 'meson.build'), 'w') as file:
+            file.write('project(\'' + artifact + '\', \'java\')\n')
+            file.write('javac = meson.get_compiler(\'java\')\n')
+            #file.write(artifact + '_dep = javac.find_library(\'' + jarname + '\')')
+            file.write(artifact + '_dep = declare_dependency(dependencies: javac.find_library(\'' + jarname + '\', dirs: [\'' + target_dir +'\']))')
```
Comment (ignatenkobrain) [inline mesonbuild/wrap/wrap.py]: __[Flake8]__

[E225] missing whitespace around operator

File: mesonbuild/wrap/wrap.py [hunk 3]
```diff
@@ -261,6 +265,31 @@ def get_svn(self, p):
...
+            dhash, jarfile = self.get_data(srcurl)
+
+        target_dir = os.path.join(self.subdir_root, p.get('directory'))
+        if not os.path.exists(target_dir):
+            os.mkdir(target_dir)
+        shutil.copyfile(jarfile, os.path.join(target_dir, jarname))
+        with open(os.path.join(target_dir, 'meson.build'), 'w') as file:
+            file.write('project(\'' + artifact + '\', \'java\')\n')
+            file.write('javac = meson.get_compiler(\'java\')\n')
+            #file.write(artifact + '_dep = javac.find_library(\'' + jarname + '\')')
```
Comment (ignatenkobrain) [inline mesonbuild/wrap/wrap.py]: __[Flake8]__

[E265] block comment should start with '# '

File: mesonbuild/wrap/wrap.py [hunk 4]
```diff
@@ -261,6 +265,30 @@ def get_svn(self, p):
...
+            dhash, jarfile = self.get_data(srcurl)
+
+        target_dir = os.path.join(self.subdir_root, p.get('directory'))
+        if not os.path.exists(target_dir):
+            os.mkdir(target_dir)
+        shutil.copyfile(jarfile, os.path.join(target_dir, jarname))
+        with open(os.path.join(target_dir, 'meson.build'), 'w') as file:
+            file.write('project(\'' + artifact + '\', \'java\')\n')
+            file.write('javac = meson.get_compiler(\'java\')\n')
+            file.write(artifact + '_dep = declare_dependency(dependencies: javac.find_library(\'' + jarname + '\', dirs: [\'' + target_dir +'\']))')
```
Comment (ignatenkobrain) [inline mesonbuild/wrap/wrap.py]: __[Flake8]__

[E225] missing whitespace around operator

File: mesonbuild/compilers/java.py [hunk 5]
```diff
@@ -113,3 +113,10 @@ def sanity_check(self, work_dir, environment):
 
     def needs_static_linker(self):
         return False
+
+    def find_library(self, libname, env, extra_dirs):
+        for d in extra_dirs:
+            trial = os.path.join(d, libname)
+            if os.path.isfile(trial):
+                return ['-cp', trial]
+        return None
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/java.py]: __[Flake8]__

[W292] no newline at end of file

**Comment by jpakkane** [comment]
There are two different things in this pr, maven and jar linking. Each one should be in its own pull request for clarity.

All new functionality must come with tests. In the linking case there would need to be a test that creates a jar target (during the test, we don't ship prebuilt binaries), and then a project that tries to find it and link to it. Judging whether changes in code make sense is almost impossible without unit tests that demonstrate how the feature is intended to be used.

**Comment by ignatenkobrain** [comment]
Flake8 detected 9 issues on 0873dc1f6eea8e5202a05ecf78a32cd099ca76b5
Visit https://sideci.com/gh/19784232/pull_requests/3732 to review the issues.

**Pull request closed without merging**